pylint/checkers/refactoring.py

```diff
@@ -459,15 +463,69 @@ class RecommandationChecker(checkers.BaseChecker):
                       'Emitted when the keys of a dictionary are iterated through the .keys() '
                       'method. It is enough to just iterate through the dictionary itself, as '
                       'in "for key in dictionary".'),
+            'C0206': ('Consider iterating the file directly instead of calling .%s()',
+                      'consider-iterating-file',
+                      'Emitted when the lines of a file are iterated through the .readlines() '
+                      'or .xreadlines() method. It is enough to just iterate through the file '
+                      'itself, as in "for line in file".'),
            }
 
     @staticmethod
+    def _is_in_iterating_context(node):
+        statement = node.statement()
+        if isinstance(statement, (astroid.Expr, astroid.Assign)):
+            statement = statement.value
+        return isinstance(statement, astroid.For) or utils.is_comprehension(statement)
+
+    @staticmethod
+    def _safe_lookup_assign_stmt(node):
+        """Given Name node, finds corresponding AssignName node"""
+        _, assign_stmts = node.lookup(node.name)
+        if len(assign_stmts) != 1:
+            return None  # looked up value may be ambiguous or uninferable
+        return assign_stmts[0]
+
+    @staticmethod
+    def _find_assigned_value(assign_stmt):
+        """Given AssignName node, finds node representing assigned value"""
+        assign_type = assign_stmt.assign_type()
+        if isinstance(assign_type, astroid.Assign):
+            return assign_type.value
+        elif isinstance(assign_type, astroid.With):
+            for value, name in assign_type.items:
+                if name == assign_stmt:
+                    return value
+        return None
+
+    @staticmethod
     def _is_builtin(node, function):
         inferred = utils.safe_infer(node)
         if not inferred:
             return False
         return utils.is_builtin_object(inferred) and inferred.name == function
 
+    @utils.check_messages('consider-iterating-file')
+    def visit_attribute(self, node):
+        if node.attrname not in FILE_READLINES_METHODS:
+            return
+
+        # we can't actually infer open() return value, so we manually check for open() calls
+        # TODO: implement inference of file (py2) / IOBase (py3) in astroid
+        open_call = None
```

Comment:
I'd rather go with this route first, before integrating this pull request, since it will simplify the implementation.